File: src/models/scheduled_occurrence.ts
export const ScheduledOccurrenceStatus = {
    Pending: 1,
    Confirmed: 2,
    Dismissed: 3
} as const;

export interface ScheduledOccurrenceSnapshotResponse {
    name: string;
    type: number;
    categoryId: string;
    sourceAccountId: string;
    destinationAccountId: string;
    sourceAmount: number;
    destinationAmount: number;
    utcOffset: number;
    hideAmount: boolean;
    tagIds: string[];
    comment: string;
}

export interface ScheduledOccurrenceInfoResponse {
    templateId: string;
    scheduledUnixTime: number;
    status: number;
    transactionId: string;
    snapshot: ScheduledOccurrenceSnapshotResponse;
}

export interface ScheduledOccurrenceActionRequest {
    templateId: string;
    scheduledUnixTime: number;
}

/**
 * One due schedule waiting for user confirmation. Confirming posts the
 * captured snapshot exactly once on the server; the occurrence itself never
 * changes an account balance.
 */
export class ScheduledOccurrence {
    readonly templateId: string;
    readonly scheduledUnixTime: number;
    readonly status: number;
    readonly transactionId: string;
    readonly name: string;
    readonly type: number;
    readonly categoryId: string;
    readonly sourceAccountId: string;
    readonly destinationAccountId: string;
    readonly sourceAmount: number;
    readonly destinationAmount: number;
    readonly utcOffset: number;
    readonly hideAmount: boolean;
    readonly comment: string;

    private constructor(info: ScheduledOccurrenceInfoResponse) {
        this.templateId = info.templateId;
        this.scheduledUnixTime = info.scheduledUnixTime;
        this.status = info.status;
        this.transactionId = info.transactionId;
        this.name = info.snapshot.name;
        this.type = info.snapshot.type;
        this.categoryId = info.snapshot.categoryId;
        this.sourceAccountId = info.snapshot.sourceAccountId;
        this.destinationAccountId = info.snapshot.destinationAccountId;
        this.sourceAmount = info.snapshot.sourceAmount;
        this.destinationAmount = info.snapshot.destinationAmount;
        this.utcOffset = info.snapshot.utcOffset;
        this.hideAmount = info.snapshot.hideAmount;
        this.comment = info.snapshot.comment;
    }

    static of(info: ScheduledOccurrenceInfoResponse): ScheduledOccurrence {
        if (!info || !info.templateId || info.templateId === '0') {
            throw new Error('invalid scheduled occurrence');
        }
        if (!info.scheduledUnixTime || info.scheduledUnixTime <= 0) {
            throw new Error('invalid scheduled occurrence time');
        }
        if (info.status !== ScheduledOccurrenceStatus.Pending &&
            info.status !== ScheduledOccurrenceStatus.Confirmed &&
            info.status !== ScheduledOccurrenceStatus.Dismissed) {
            throw new Error('invalid scheduled occurrence status');
        }
        const snapshot = info.snapshot;
        if (!snapshot || !snapshot.categoryId || snapshot.categoryId === '0' ||
            !snapshot.sourceAccountId || snapshot.sourceAccountId === '0') {
            throw new Error('incomplete scheduled occurrence snapshot');
        }
        return new ScheduledOccurrence(info);
    }

    get isPending(): boolean {
        return this.status === ScheduledOccurrenceStatus.Pending;
    }

    get isDismissed(): boolean {
        return this.status === ScheduledOccurrenceStatus.Dismissed;
    }

    get actionRequest(): ScheduledOccurrenceActionRequest {
        return {
            templateId: this.templateId,
            scheduledUnixTime: this.scheduledUnixTime
        };
    }
}

File: pkg/models/scheduled_occurrence.go
package models

import "fmt"

// ScheduledOccurrence is a durable review item, not a posted transaction.
// The composite primary key makes dispatch of a given occurrence idempotent.
type ScheduledOccurrence struct {
	Uid               int64                     `xorm:"PK NOT NULL"`
	TemplateId        int64                     `xorm:"PK NOT NULL"`
	ScheduledUnixTime int64                     `xorm:"PK NOT NULL"`
	Status            ScheduledOccurrenceStatus `xorm:"NOT NULL"`
	SnapshotJSON      string                    `xorm:"TEXT NOT NULL"`
	TransactionId     int64
	CreatedUnixTime   int64
	UpdatedUnixTime   int64
}

type ScheduledOccurrenceStatus byte

const (
	ScheduledOccurrencePending   ScheduledOccurrenceStatus = 1
	ScheduledOccurrenceConfirmed ScheduledOccurrenceStatus = 2
	ScheduledOccurrenceDismissed ScheduledOccurrenceStatus = 3
)

// ScheduledOccurrenceSnapshot preserves monetary values and IDs without float conversion.
type ScheduledOccurrenceSnapshot struct {
	Name                 string          `json:"name"`
	Type                 TransactionType `json:"type"`
	CategoryId           int64           `json:"categoryId,string"`
	SourceAccountId      int64           `json:"sourceAccountId,string"`
	DestinationAccountId int64           `json:"destinationAccountId,string"`
	SourceAmount         int64           `json:"sourceAmount"`
	DestinationAmount    int64           `json:"destinationAmount"`
	UtcOffset            int16           `json:"utcOffset"`
	HideAmount           bool            `json:"hideAmount"`
	TagIds               []string        `json:"tagIds"`
	Comment              string          `json:"comment"`
}

// Validate rejects incomplete persisted snapshots before they reach the ledger.
func (s ScheduledOccurrenceSnapshot) Validate() error {
	const maximumAmount = int64(9999999999999)
	if s.Type < TRANSACTION_TYPE_INCOME || s.Type > TRANSACTION_TYPE_TRANSFER || s.SourceAccountId <= 0 || s.CategoryId <= 0 ||
		s.SourceAmount < -maximumAmount || s.SourceAmount > maximumAmount ||
		s.DestinationAmount < -maximumAmount || s.DestinationAmount > maximumAmount || s.UtcOffset < -720 || s.UtcOffset > 840 || len(s.TagIds) > MaximumTagsCountOfTransaction {
		return fmt.Errorf("invalid scheduled transaction snapshot")
	}
	if s.Type == TRANSACTION_TYPE_TRANSFER {
		if s.DestinationAccountId <= 0 || s.DestinationAccountId == s.SourceAccountId || s.SourceAmount < 0 || s.DestinationAmount < 0 {
			return fmt.Errorf("invalid scheduled transfer")
		}
	} else if s.DestinationAccountId != 0 || s.DestinationAmount != 0 {
		return fmt.Errorf("non-transfer snapshot has destination fields")
	}
	return nil
}
